main/sdr.go

```diff
@@ -319,21 +314,164 @@ func sdrKill() {
+	//         UAT    ES    id 1 -> ES  second pass loop 1, id 0 -> Unconfigured
+	//         ES     ES    id 0 -> ES  first  pass loop 1, id 1 -> unconfigured
+	//         UAT    UAT   id 0 -> Unconfigured          , id 1 -> Unconfigured
+
+	// es_enabled == false and uat_enabled == true
+	//
+	// ID      0      1
+	// -------------------------------------------------------------------------
+	//         A      A     id 0 -> UAT first  pass loop 2, id 1 -> Unconfigured
+	//         A      ES	id 0 -> UAT first  pass loop 2, id 1 -> Unconfigured
+	//         ES     A     id 1 -> UAT second pass loop 2, id 0 -> Unconfigured
+	//         A      UAT   id 1 -> UAT second pass loop 1, id 0 -> Unconfigured
+	//         UAT    A     id 0 -> UAT first  pass loop 1, id 1 -> Unconfigured
+	//         ES     UAT   id 1 -> UAT second pass loop 1, id 0 -> Unconfigured
+	//         UAT    ES    id 0 -> UAT first  pass loop 1, id 1 -> Unconfigured
+	//         ES     ES    id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         UAT    UAT   id 0 -> UAT first  pass loop 1, id 1 -> unconfigured
+
+	// es_enabled == false and uat_enabled == false
+	//
+	// ID      0      1
+	// -------------------------------------------------------------------------
+	//         A      A     id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         A      ES	id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         ES     A     id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         A      UAT   id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         UAT    A     id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         ES     UAT   id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         UAT    ES    id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         ES     ES    id 0 -> Unconfigured          , id 1 -> Unconfigured
+	//         UAT    UAT   id 0 -> Unconfigured          , id 1 -> Unconfigured
+
+	// once the tagged dongles have been assigned explicitly range over
+	// the remaining IDs and assign then to any anonymous dongles
+	unusedIDs := make([]int, 0, count)
+
+	// loop 1: assign tagged dongles
+	for i := 0; i < count; i++ {
+		_, _, s, err := rtl.GetDeviceUsbStrings(i)
+		if err == nil {
+			if uat_enabled && UATDev == nil && rUAT.hasID(s) {
+				createUATDev(i, s, true)
+			} else if es_enabled && ESDev == nil && rES.hasID(s) {
+				createESDev(i, s, true)
+			} else {
+				unusedIDs = append(unusedIDs, i)
+			}
+		} else {
+			log.Printf("rtl.GetDeviceUsbStrings id %d: %s\n", i, err)
+		}
 	}
 
+	// loop 2; assign anonymous dongles
+	for _, v := range unusedIDs {
+		_, _, s, err := rtl.GetDeviceUsbStrings(v)
+		if err == nil {
+			if uat_enabled && UATDev == nil {
+				createUATDev(v, s, false)
+			} else if es_enabled && ESDev == nil {
+				createESDev(v, s, false)
+			}
+		} else {
+			log.Printf("rtl.GetDeviceUsbStrings id %d: %s\n", v, err)
+		}
+	}
+}
```

Comment:
On Mon, Feb 22, 2016 at 11:39 AM, yihchun <notifications@github.com> wrote:

> In main/sdr.go
> <https://github.com/cyoung/stratux/pull/279#discussion_r53661410>:
>
> >  	}
> >
> > +	// loop 2; assign anonymous dongles
> > +	for _, v := range unusedIDs {
> > +		_, _, s, err := rtl.GetDeviceUsbStrings(v)
> > +		if err == nil {
> > +			if uat_enabled && UATDev == nil {
> > +				createUATDev(v, s, false)
> > +			} else if es_enabled && ESDev == nil {
> > +				createESDev(v, s, false)
> > +			}
> > +		} else {
> > +			log.Printf("rtl.GetDeviceUsbStrings id %d: %s\n", v, err)
> > +		}
> > +	}
> > +}
>
>
one caveat is that we only support up to two dongles, the first two seen by
librtlsdr (device id 0 and 1) :)

        // support two and only two dongles
        if count > 2 {
            count = 2
        }

> Looks great but I think your big comment block is incorrect; since loop 2
> doesn't check, for example,
> if uat_enabled && UATDev == nil && !rES.hasID(s)
> loop 2 will automatically override (ignore?) any dual-assignments. In my
> opinion, ideal behavior would include such a check in loop 2 and then
> override
>
I'm not sure I understand.

Once loop 1 finishes the only remaining *enabled* dongles *must* be
anonymous, therefore, in loop 2, rES.hasID(s) and rUAT.hasID(s) could never
be true for an enabled and as yet unassigned device.

unusedIDs contains no IDs prior to loop 1 and device ID 0 is only appended
to unusedIDs when said dongle has neither a 1090 or 978 serial, similar for
device ID 1. Therefore, when loop 2 starts we only range over device IDs
that weren't consumed in loop 1.